src/devices/rpLidar2/sdk_original/sdk/src/arch/win32/net_serial.cpp

```diff
@@ -0,0 +1,358 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "sdkcommon.h"
+#include "net_serial.h"
+
+namespace rp{ namespace arch{ namespace net{
+
+raw_serial::raw_serial()
+    : rp::hal::serial_rxtx()
+    , _serial_handle(NULL)
+    , _baudrate(0)
+    , _flags(0)
+{
+    _init();
+}
+
+raw_serial::~raw_serial()
+{
+    close();
+
+    CloseHandle(_ro.hEvent);
+    CloseHandle(_wo.hEvent);
+    CloseHandle(_wait_o.hEvent);
+}
+
+bool raw_serial::open()
+{
+    return open(_portName, _baudrate, _flags);
+}
+
+bool raw_serial::bind(const char * portname, _u32 baudrate, _u32 flags)
+{   
+    strncpy(_portName, portname, sizeof(_portName));
+    _baudrate = baudrate;
+    _flags    = flags;
+    return true;
+}
+
+bool raw_serial::open(const char * portname, _u32 baudrate, _u32 flags)
+{
+    if (isOpened()) close();
+    
+    _serial_handle = CreateFile(
+        portname,
+        GENERIC_READ | GENERIC_WRITE,
+        0,
+        NULL,
+        OPEN_EXISTING,
+        FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OVERLAPPED,
+        NULL
+        );
+
+    if (_serial_handle == INVALID_HANDLE_VALUE) return false;
+
+    if (!SetupComm(_serial_handle, SERIAL_RX_BUFFER_SIZE, SERIAL_TX_BUFFER_SIZE))
+    {
+        close();
+        return false;
+    }
+    
+    _dcb.BaudRate = baudrate;
+    _dcb.ByteSize = 8;
+    _dcb.Parity   = NOPARITY;
+    _dcb.StopBits = ONESTOPBIT;
+    _dcb.fDtrControl = DTR_CONTROL_ENABLE;
+
+    if (!SetCommState(_serial_handle, &_dcb))
+    {
+        close();
+        return false;
+    }
+
+    if (!SetCommTimeouts(_serial_handle, &_co))
+    {
+        close();
+        return false;
+    }
+
+    if (!SetCommMask(_serial_handle, EV_RXCHAR | EV_ERR ))
+    {
+        close();
+        return false;
+    }
+
+    if (!PurgeComm(_serial_handle, PURGE_TXABORT | PURGE_RXABORT | PURGE_TXCLEAR | PURGE_RXCLEAR ))
+    {
+        close();
+        return false;
+    }
+
+    Sleep(30); 
+    _is_serial_opened = true;
+
+    //Clear the DTR bit set DTR=high
+    clearDTR();
+
+    return true;
+}
+
+void raw_serial::close()
+{
+    SetCommMask(_serial_handle, 0);
+    ResetEvent(_wait_o.hEvent);
+
+    CloseHandle(_serial_handle);
+    _serial_handle = INVALID_HANDLE_VALUE;
+    
+    _is_serial_opened = false;
+}
+
+int raw_serial::senddata(const unsigned char * data, size_t size)
+{
+    DWORD    error;
+    DWORD w_len = 0, o_len = -1;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'o_len' is assigned a value that is never used.](https://app.codacy.com/manual/robotology/yarp/pullRequest?prid=5162834)